src/pgen/cluster.cpp

```diff
@@ -405,4 +435,107 @@ void ProblemGenerator(MeshBlock *pmb, parthenon::ParameterInput *pin) {
   } // END if(hydro_pkg->Param<Fluid>("fluid") == Fluid::glmmhd)
 }
 
+void UserWorkBeforeOutput(MeshBlock *pmb, ParameterInput *pin) {
+  // get hydro
+  auto pkg = pmb->packages.Get("Hydro");
+  const Real gam = pin->GetReal("hydro", "gamma");
+  const Real gm1 = (gam - 1.0);
+
+  // get prim vars
+  auto &data = pmb->meshblock_data.Get();
+  auto const &prim = data->Get("prim").data;
+
+  // get derived fields
+  auto &log10_radius = data->Get("log10_cell_radius").data;
+  auto &entropy = data->Get("entropy").data;
+  auto &mach_sonic = data->Get("mach_sonic").data;
+  auto &temperature = data->Get("temperature").data;
+
+  // for computing temperature from primitives
+  auto mbar_over_kb = pkg->Param<Real>("mbar_over_kb");
+
+  // fill derived vars (*including ghost cells*)
+  auto &coords = pmb->coords;
+  IndexRange ib = pmb->cellbounds.GetBoundsI(IndexDomain::entire);
+  IndexRange jb = pmb->cellbounds.GetBoundsJ(IndexDomain::entire);
+  IndexRange kb = pmb->cellbounds.GetBoundsK(IndexDomain::entire);
+
+  pmb->par_for(
+      "Cluster::UserWorkBeforeOutput", kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int k, const int j, const int i) {
+        // get gas properties
+        const Real rho = prim(IDN, k, j, i);
+        const Real v1 = prim(IV1, k, j, i);
+        const Real v2 = prim(IV2, k, j, i);
+        const Real v3 = prim(IV3, k, j, i);
+        const Real P = prim(IPR, k, j, i);
+
+        // compute radius
+        const Real x = coords.Xc<1>(i);
+        const Real y = coords.Xc<2>(j);
+        const Real z = coords.Xc<3>(k);
+        const Real r2 = SQR(x) + SQR(y) + SQR(z);
+        log10_radius(k, j, i) = 0.5 * std::log10(r2);
+
+        // compute entropy
+        const Real K = P / std::pow(rho, gam);
+        entropy(k, j, i) = K;
+
+        const Real v_mag = std::sqrt(SQR(v1) + SQR(v2) + SQR(v3));
+        const Real c_s = std::sqrt(gam * P / rho); // ideal gas EOS
+        const Real M_s = v_mag / c_s;
+        mach_sonic(k, j, i) = M_s;
+
+        // compute temperature
+        temperature(k, j, i) = mbar_over_kb * P / rho;
+      });
+  if (pkg->Param<Cooling>("enable_cooling") == Cooling::tabular) {
+    auto &cooling_time = data->Get("cooling_time").data;
+
+    // get cooling function
+    const cooling::TabularCooling &tabular_cooling =
+        pkg->Param<cooling::TabularCooling>("tabular_cooling");
+    const auto cooling_table_obj = tabular_cooling.GetCoolingTableObj();
+
+    pmb->par_for(
+        "Cluster::UserWorkBeforeOutput::CoolingTime", kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+        KOKKOS_LAMBDA(const int k, const int j, const int i) {
+          // get gas properties
+          const Real rho = prim(IDN, k, j, i);
+          const Real P = prim(IPR, k, j, i);
+
+          // compute cooling time
+          const Real eint = P / (rho * gm1);
+          const Real edot = cooling_table_obj.DeDt(eint, rho);
+          cooling_time(k, j, i) = (edot != 0) ? eint / edot : NAN;
```

Comment:
This specific line causes issues on Frontier:
```
Memory access fault by GPU node-10 (Agent handle: 0x2c97530) on address (nil). Reason: Page not present or supervisor privilege.
Memory access fault by GPU node-6 (Agent handle: 0x2c97530) on address (nil). Reason: Page not present or supervisor privilege.
Memory access fault by GPU node-4 (Agent handle: 0x2c97530) on address (nil). Reason: Page not present or supervisor privilege.
Memory access fault by GPU node-8 (Agent handle: 0x2c97530) on address (nil). Reason: Unknown.
```

Commenting it out lets the simulation run. What's frustrating is `cooling_table_obj.DeDt` seemingly works fine elsewhere when running on Frontier

Reply:
Does `rocgdb` work on Frontier? That *might* provide more detail (e.g., https://www.olcf.ornl.gov/wp-content/uploads/2021/04/rocgdb_hipmath_ornl_2021_v2.pdf)

Reply:
FYI I tried also a full new stack with rom 5.4.3 (as googling pointed to some potential compiler issues), but this didn't fix it.
I'm also not sure why it's not working but I'm a little bit suspicious/afraid that this may point to a deeper (real) issue on hip with the new object.
In other words, I'd prefer if we hold of merging for now until we understand where that issue comes from.

Reply:
Could we split the PR into cooling time vs. the rest of the fields? We could at least get the other fields and test that the analysis work in production.

Reply:
I'd be in favor of splitting up this PR into two. I'll try to do that this weekend. I'll take a stab at using the old clunky cooling rate function to still include the cooling time.

Very disappointing though, I liked the new refactor.

Reply:
The refactor is undone, saving the cooling time apparently works now on Frontier

Reply:
Did this turn out to be a compiler bug or something like that?